Introduce UntypedQuery, MutationQuery, UntypedNativeQuery, MutationNa…

sebersole opened the pull request:
Introduce UntypedQuery, MutationQuery, UntypedNativeQuery, MutationNa…tiveQuery, SelectionQuery, JpaQuery, etc;
HHH-14855 - Support passing a String representation of EntityGraph we can parse through the graph-language parser

gavinking commented:
@sebersole where do you wanna discuss this? Here on the PR, or in a discussion, or?

yrodiere commented on hibernate-core/src/main/java/org/hibernate/query/QueryProducer.java:
```diff
@@ -33,11 +33,22 @@
-	 * @deprecated use {@link #createQuery(String, Class)} or {@link #createStatement(String)}
+	 *
+	 * @deprecated use {@link #createQuery(String, Class)},
+	 * {@link #createUntypedQuery}, {@link #createMutationQuery(String)}
+	 * or {@link #createMutationQuery} depending on intention
```
The second `{@link #createMutationQuery}` doesn't mention method parameters. Is this a mistake?

sebersole replied:
Its just an extra link, a mistake in it being there period.  There is only one form of `#createMutationQuery`

yrodiere commented on hibernate-core/src/main/java/org/hibernate/query/CommonQueryContract.java:
```diff
@@ -20,6 +28,424 @@
  * @author Gavin King
  */
 public interface CommonQueryContract {
```
I get why it's necessary to have a common interface and then override all methods in sub-interfaces (to return the sub-interface instead of the common interface).

However, as you know, there's a real risk of forgetting to add overrides when we add a new method to the common interface. Well, it's not even a risk, it's a certainty: we will forget at some point.

On projects using Maven, I would have used [JQAssistant](https://jqassistant.org/) to check that all sub-interfaces of `CommonQueryContract` override all methods that return a `CommonQueryContract` with the proper return type.
But with Gradle, you'd have to use [this unofficial plugin](https://github.com/kontext-e/jqassistant-gradle-plugin), and I'm not sure what it's worth. Plus, setting this up may be complicated.

Still, using this inheritance strategy without automated checks makes me uneasy...

gavinking commented:
Well, OK, looks like @yrodiere has started giving feedback here, so I'll post my comments here too, which goes against my preference (GitHub pull requests are a terrible place for discussions, IMO) but whatever. @sebersole if you prefer to move to the Discussions, I'll delete this comment and repost it there.

So, my overall feeling on this is I'm pretty "meh", but I could be sold on it if you can point to some really concrete benefit to the user.

My initial reaction, without wanting to dig myself too deep into any ratholes is the following:

- Conceptually, and _in principle_, this design is pretty sound: each of these new interfaces is a well-defined type, 
- except&mdash;at least arguably&mdash;for `UntypedQuery`, which is really just a `Query<Object>`, but
- on the other hand, it’s a significant escalation in the complexity that's visible to the user, and 
- it's not _yet_ crystal clear to me what benefit the user gets out of this.

Drilling down a little:

- What really makes me doubt the value of this change is it doesn’t really improve type safety (much?) that I can see: `MutationQuery` still has `getResultList()`, and `Query` still has `executeUpdate()`, which to me pretty much undermines the whole purpose of having special types to represent these things. What is the actual difference between `MutationQuery` and `Query<Void>`? Which operations do they not have in common?
- Such a deep hierarchy makes it quite a lot harder to navigate to the Javadoc’d declaration of operations, with so much overriding that just changes return types. Perhaps your initial reaction is to sorta pooh-pooh this objection&mdash;and in a completely different context, that might be my reaction too&mdash;but this is the kind of thing that actually impacts me quite a lot when I sit down to write code.
- I'm guessing that the motivation for introducing `UntypedQuery` is that we'll use it _in future_ as the return type of the methods inherited from JPA that we marked `@Deprecated` in this release; if that's the case, then fine, I get that, and I'm not against it, but then it's not clear to me that the best time to add it is _right now_. Could we wait and introduce it later when the time is riper?

Now some much smaller objections that can I guess be fixed pretty easily:

- `UntypedQuery` is defined in terms of raw types: `getResultList()` returns a raw list instead of a `List<Object>`. I don't see why it can't return `List<Object>` to protect the user from getting a compiler warning that they then have to suppress.
- `CommonQueryContract` still declares some operations that I believe aren’t relevant to “mutation” queries.
- I’m not particularly happy to see `@SuppressWarnings("rawtypes")` migrate back up to the interface level - I worked hard to make sure it was always applied where the raw type is instantiated, instead of just sucking in all operations of the type.
- Javadoc like “`{@link jakarta.persistence.Query} override`” is a bad idea, it adds no extra information that IntelliJ doesn’t give me automatically, and it prevents IntelliJ from displaying the Javadoc of the overridden method when I hover over an invocation.

Again, I don't want to paint myself into a corner here: I'm not implacably against this change, it's not a terrible idea, I'm just on the fence, and I want to see why it's a real improvement.

gavinking commented on hibernate-core/src/main/java/org/hibernate/query/Query.java:
```diff
@@ -37,63 +37,41 @@
-	 * @return this - for method chaining
-	 */
-	Query<R> applyGraph(@SuppressWarnings("rawtypes") RootGraph graph, GraphSemantic semantic);
+public interface Query<R> extends UntypedQuery, JpaQuery, TypedQuery<R> {
```
> I believe you added `UntypedQuery` because you wanted to stress that a query is untyped, but Java (for better or worse) already provides a way to do that: raw types.

@yrodiere it's a little more subtle than that. The issue is that a raw `Query` not only erases the (conceptually covariant) type parameter that occurs in the return types of the operations of `Query`, but, unfortunately, also all contravariant type arguments that occur in parameter types of the operations. 

This means, in our context, that you get all sorts of nasty and completely bogus warnings when you call operations of `Query`, including `setParameter()` and `addEntity()`&mdash;you're at risk of a warning anytime you call any operation with a generic type instantiation in its schema.

So it's not quite as simple as "oh, this is just the same as a raw type".

yrodiere replied:
Ok, but wouldn't this solve the problem just as well?

> I would suggest to make `SelectionQuery` parameterized (`SelectionQuery<R>`), add parameters whevever it makes sense (e.g. `Query<R> extends SelectionQuery<R>`), and use `SelectionQuery<?>` for legacy, untyped queries (e.g. `SelectionQuery<?> createdUntypedQuery(...)`. That would simplify your inheritance tree, and we really need all the simplifying we can get here.

I.e. making `SelectionQuery` parameterized and using `SelectionQuery<?>` (or `SelectionQuery<Object>` like you suggested elsewhere) in APIs. No raw types, no problems.

gavinking replied:
Yeah, apologies: I was going to let Steve respond to that, but I read it and it didn't sound crazy or obviously unworkable to me. But without actually trying it out I can't say for sure if it works or not.

sebersole replied:
> This looks weird: `Query<R> extends UntypedQuery, [...] TypedQuery<R>`.

Hm, not sure where you are seeing that.  I have:

```
public interface SelectionQuery extends CommonQueryContract {
    ...
}

public interface UntypedQuery extends SelectionQuery {
    ...
}
```

> 
> I believe you added `UntypedQuery` because you wanted to stress that a query is untyped, but Java (for better or worse) already provides a way to do that: raw types.
> 
> Plus, IMO a contract only makes sense if it offers operations. `UntypedQuery` is basically the same as `SelectionQuery`, but with a different name. Anywhere someone would use `UntypedQuery`, they could use `SelectionQuery` instead. So... I'm not sure it makes sense to ever use `UntypedQuery` in an application? Unless they really want to make sure their query is untyped, but why would they want that?

I'm ok with dropping `UntypedQuery` altogether in favor of `SelectionQuery`.

Initially `UntypedQuery` hooked in to the JPA Query hierarchy whereas `SelectionQuery` did not.  Forgot I changed that.  The intention being that users could treat `UntypedQuery` as `jakarta.persistence.Query`.  But you are right, once I dropped `UntypedQuery extends jakarta.persistence.Query` it became a moot point.

Ultimately I guess it comes down to how we want to handle `Query#createQuery(String)`


> I would suggest to make `SelectionQuery` parameterized (`SelectionQuery<R>`), add parameters whevever it makes sense (e.g. `Query<R> extends SelectionQuery<R>`), and use `SelectionQuery<?>` for legacy, untyped queries (e.g. `SelectionQuery<?> createdUntypedQuery(...)`. That would simplify your inheritance tree, and we really need all the simplifying we can get here.

But that's not an untyped (to me == raw) query really

sebersole replied:
> I believe you added `UntypedQuery` because you wanted to stress that a query is untyped, but Java (for better or worse) already provides a way to do that: raw types.

No, I did it to somewhat mirror JPA's own Query hierarchy - just with (imo ofc) better naming

sebersole replied:
> we really need all the simplifying we can get here.

Meh - i don't think its all that complicated really.  Yes the covariant stuff blows, but well... Java

yrodiere replied:
> > This looks weird: `Query<R> extends UntypedQuery, [...] TypedQuery<R>`.
> 
> Hm, not sure where you are seeing that. I have:

See above my comment. It's in the line I annotated.



> I'm ok with dropping `UntypedQuery` altogether in favor of `SelectionQuery`.

:+1: 

> But that's not an untyped (to me == raw) query really

Sure, it's not a raw type. I'd say it's still untyped since we don't promise that the returned hits will have a specific type, but YMMV.

In any case, I'd argue it's better than a raw query type:

* It accurately expresses the idea that "we don't know what type of hits this query returns" (just like a raw type would)
* It avoids the kind of problems we get from raw types (see [Gavin's comment above](https://github.com/hibernate/hibernate-orm/pull/4642#discussion_r785806689))
* It provides more useful compile-time checks to users when they incorrectly the query returns a specific type: for example `List<MyEntity> list = session.createUntypedQuery(...).getResultList();` will fail to compile, instead of just raising a warning. People will need to acknowledge that their query returns... untyped... entities and handle that.

Now, if you really want raw types for some reason, I guess you can use a raw `SelectionQuery`... That could work too, with the caveats mentioned by Gavin in his comment.

yrodiere replied:
AFAICS this was addressed by using non-raw return types in `SelectionQuery`, but `SelectionQuery` itself still isn't generic. Well, I guess time will tell if this was a good idea.

gavinking commented:
To be clear, @sebersole, I would be OK with all this if it weren't for the fact that in the end all flavors of `Query` wind up inheriting `getResultList()` and `executeUpdate()` from JPA's `Query`.

It's the fact that everything ultimately descends from JPA `Query`&mdash;which does not distinguish between mutation queries and query queries&mdash;which makes me question the whole approach.

If it weren't for that, y'know, I would be totally on board with all this and might even have proposed it myself.

gavinking commented:
> `UntypedQuery` is defined in terms of raw types: `getResultList()` returns a raw list instead of a `List<Object>`. I don't see why it can't return `List<Object>` to protect the user from getting a compiler warning that they then have to suppress.

Oh, excuse me, my bad, it's because `Query` extends `UntypedQuery` rather than the other way round.

(PS: I'm not sure I think that's the best option, but I need to look more again at what you've done and think some more. Naively, I would have written `UntypedQuery extends Query<Object>`, but I know there's some subtleties going on here.)

So, anyway, assuming your design, I would say `UntypedQuery.getResultList()` should have the return type `List<?>` rather than the raw type `List`.

sebersole commented:
> So, my overall feeling on this is I'm pretty "meh", but I could be sold on it if you can point to some really concrete benefit to the user.

I like that users get earlier feedback about the query not being the correct type (creating the Query reference) rather than waiting for the execution.  

I like that it also limits the available options... Unlike both Hibernate's and JPA's Query expose operations pertaining to SelectionQuery (`#list`, ...) and MutationQuery (`#executeUpdate`).  Splitting them out is also nice I think

sebersole commented:
> I would say UntypedQuery.getResultList() should have the return type List<?> rather than the raw type List

`List<?>` is not untyped though.  Nor does it match typing-wise with JPA's `Query` (as distinct from its `TypedQuery`).  To me, untyped here == raw; otherwise it is typed, just unbounded

sebersole commented:
> To be clear, @sebersole, I would be OK with all this if it weren't for the fact that in the end all flavors of `Query` wind up inheriting `getResultList()` and `executeUpdate()` from JPA's `Query`.

That's just not true.  *Query* yes, but that is completely reasonable for exactly the reason you mentioned - they in fact are part of our extension to JPA.  Not to mention, its not even just JPA - our own Query interface did and does the same exact thing, even before I folded the Hibernate and JPA Query into one.  Nothing new (nor bad) there.

However, neither `MutationQuery` nor `SelectionQuery` do - those are "well defined" in that they only expose the relevant operations

sebersole commented:
Look, I am completely fine with not applying this.  But then I do not want to hear anymore about `Session#createQuery` and it returning an unbounded, typed result.  That's what started this entire thing btw - y'all complained about the return there.  

We can go down this path (with some tweaks maybe) or just leave it as is and move on.

gavinking commented:
> However, neither `MutationQuery` nor `SelectionQuery` do - those are "well defined" in that they only expose the relevant operations

In what I have, `MutationQuery` extends `jakarta.persistence.Query`, which declares `getResultList()`.

sebersole commented:
> > However, neither `MutationQuery` nor `SelectionQuery` do - those are "well defined" in that they only expose the relevant operations
> 
> In what I'm looking at, `MutationQuery` extends `jakarta.persistence.Query`, which declares `getResultList()`.

Ah, right - I had added that back so the forms from JPA accepting criteria trees could return a `MutationQuery`.  I removed `MutationQuery` extending from JPA Query, and just have those methods continue to return query.  I added these new methods:

```
MutationQuery createMutationQuery(@SuppressWarnings("rawtypes") CriteriaUpdate updateQuery);

MutationQuery createMutationQuery(@SuppressWarnings("rawtypes") CriteriaDelete deleteQuery);
```

Pushing in a few minutes

gavinking commented:
> Pushing in a few minutes

OK, give me a chance to take another look (a bit later today).

yrodiere commented on hibernate-core/src/main/java/org/hibernate/query/Query.java:
```diff
@@ -37,63 +37,43 @@
-	 * @return the query string.
-	 */
-	String getQueryString();
+public interface Query<R> extends SelectionQuery, JpaQuery, TypedQuery<R> {
```
I don't get why `SelectionQuery` isn't `SelectionQuery<R>` instead?

I'm not sure raw types (e.g. in `getResultList()`) will bring any value to users. They'll be better off with typed results, and *when* we need untyped queries in the API, we can use `SelectionQuery<Object>` or `SelectionQuery<?>` (that's what wildcards are for). Unless there's an actual reason to use raw types that I didn't catch... ?

sebersole replied:
```
package jakarta.persistence;

public interface Query {
   ...
}
```

yrodiere replied:
Sure, but AFAICS, all type that implement `SelectionQuery` _extend_ JPA's Query type. JPA's Query type doesn't extend `SelectionQuery`. So you're free to "do better" than JPA here. For example by adding generics: `SelectionQuery<R>`. Just like you did in `org.hibernate.query.Query<R>`, or like JPA did in `TypedQuery`, really.

Or did using generics lead to problems related to contravariance of R, e.g. a method of `SelectionQuery` accepting a parameter of type `R`? I'd be surprised, but if such a problem arises, there are other solutions than just sticking to raw types.

sebersole commented:
Integrated upstream
